rust_src/src/buffers.rs

```diff
@@ -1131,6 +1131,57 @@ pub unsafe fn per_buffer_idx(offset: isize) -> isize {
     (*obj).as_fixnum_or_error() as isize
 }
 
+/// Return a list of overlays which is a copy of the overlay list
+/// LIST, but for buffer BUFFER.
+#[no_mangle]
+pub unsafe extern "C" fn copy_overlays(
+    buffer: *mut Lisp_Buffer,
+    list: *mut Lisp_Overlay,
+) -> *mut Lisp_Overlay {
+    if list == ptr::null_mut() {
+        return list;
+    }
+
+    let mut result = ptr::null_mut();
+
+    let overlays_iter = LispOverlayRef::from_ptr(list as *mut c_void)
+        .unwrap_or_else(|| panic!("Invalid overlay reference."))
+        .iter();
+
+    let _ = overlays_iter.fold(None, |mut tail: Option<LispOverlayRef>, overlay| {
```

Comment:
The start and end logic do the same work. One potential simplification is to make a function out of it. Then the body of the fold becomes:

    let start = duplicate_marker(overlay.start.as_marker_or_error());
    let end = duplicate_marker(overlay.end.as_marker_or_error());
    let mut overlay_new = build_overlay(...);
    match tail {
        Some(mut tail_ref) => tail_ref.next = overlay_new.as_mut(),
        None => result = overlay_new.as_mut(),
    }
    tail = Some(overlay_new);
    tail

Reply:
Note the update to tail is pulled out of the `match` because it is the same in both arms. `match` is more idiomatic here too.

Reply:
`duplicate_marker` could be defined as a local lambda instead of making a full function.

Reply:
Thanks. Done

Reply:
Before, the callback was too long and unwieldy. I was going to suggest using a loop instead. Your update is much easier to read. Incorporating `tail` in the fold argument is perfect.